schedule/rotation/rotation.go

```diff
@@ -32,10 +32,57 @@ func (r Rotation) shiftClock() timeutil.Clock {
 	case TypeWeekly:
 		return timeutil.NewClock(r.ShiftLength*24*7, 0)
 	default:
+		// monthly is handled separately
 		panic("unexpected rotation type")
 	}
 }
 
+func (r Rotation) monthStartTime(t time.Time, n int) time.Time {
+	if n > 10000 {
+		panic("too many iterations")
+	}
+
+	if t.After(r.Start) || t.Equal(r.Start) { // t is after start of rotation
+		next := r.Start.AddDate(0, r.ShiftLength*n, 0)
+		if next.After(t) {
+			return r.Start.AddDate(0, r.ShiftLength*(n-1), 0)
+		}
+
+		return r.monthStartTime(t, n+1)
+	}
+
+	// t is before start of rotation
+	prev := r.Start.AddDate(0, -r.ShiftLength*n, 0)
+	if prev.Before(t) {
+		return prev
+	}
+
+	return r.monthStartTime(t, n+1)
+}
```

Comment:
can we add some more context here please?

Is the month start time of what? The shift? The month itself?

Why are we multiplying by a random int always sent as `1` and returning different things based off that? A comment above these and/or StartTime would be helpful.

There is the comment "For daily and weekly rotations, start time will be the previous handoff time (from start)."

It would be useful if that comment also says why monthly is being calculated differently

Reply:
comments added,
monthStartTime/monthEndTime follow the same behavior as daily/weekly rotation in the startTime/endTime functions. The previous handoff time of the rotation starting from or active at time t is calculated for monthStartTime.
The handoff time of the rotation starting from or active at time t is calculated for monthEndTime.

Recursion was used here as the length of months vary meaning that we could not simply add/subtract the r.shiftClock() value. It made more sense to recursively calculate with the AddDate() function to accurately determine the start/end times.

Reply:
Thank you for the clarification! It should help with maintenance in the future.